primitives/core/storage/src/lib.rs

```diff
@@ -156,3 +159,198 @@ impl<'a> ChildStorageKey<'a> {
+/// Information related to a child state.
+pub enum ChildInfo<'a> {
+	Default(ChildTrie<'a>),
+}
+
+/// Owned version of `ChildInfo`.
+/// To be use in persistence layers.
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "std", derive(PartialEq, Eq, Hash, PartialOrd, Ord))]
+pub enum OwnedChildInfo {
+	Default(OwnedChildTrie),
+}
+
+impl<'a> ChildInfo<'a> {
+	/// Instantiates information for a default child trie.
+	pub const fn new_default(unique_id: &'a[u8]) -> Self {
+		ChildInfo::Default(ChildTrie {
+			root: 0,
+			data: unique_id,
+		})
+	}
+
+	/// Instantiates a owned version of this child info.
+	pub fn to_owned(&self) -> OwnedChildInfo {
+		match self {
+			ChildInfo::Default(ChildTrie { data, root })
+				=> OwnedChildInfo::Default(OwnedChildTrie {
+					root: *root,
+					data: data.to_vec(),
+				}),
+		}
+	}
+
+	/// Create child info from a linear byte packed value and a given type. 
+	pub fn resolve_child_info(child_type: u32, data: &'a[u8]) -> Option<Self> {
+		match child_type {
+			x if x == ChildType::CryptoUniqueId as u32 => Some(ChildInfo::new_default(data)),
+			x if x == ChildType::CryptoUniqueIdRoot32Api as u32 => if data.len() >= 32 {
+				Some(ChildInfo::Default(ChildTrie {
+					root: 32,
+					data,
+				}))
+			} else { None },
+			_ => None,
+		}
+	}
+
+	/// Return a single byte vector containing packed child info content and its child info type.
+	/// This can be use as input for `resolve_child_info`.
+	pub fn info(&self) -> (&[u8], u32) {
+		match self {
+			ChildInfo::Default(ChildTrie {
+				root,
+				data,
+			}) => {
+				if *root > 0 {
+					if *root == 32 {
+						return (data, ChildType::CryptoUniqueIdRoot32Api as u32);
+					}
+				}
+				(data, ChildType::CryptoUniqueId as u32)
+			},
+		}
+	}
+
+	/// Return byte sequence (keyspace) that can be use by underlying db to isolate keys.
+	/// This is a unique id of the child trie. The collision resistance of this value
+	/// depends on the type of child info use. For `ChildInfo::Default` it is and need to be.
+	pub fn keyspace(&self) -> &[u8] {
+		match self {
+			ChildInfo::Default(ChildTrie {
+				root,
+				data,
+			}) => &data[*root..],
+		}
+	}
+
+	/// Return the child reference to state if it is already known.
+	/// If it returns `None` the information will need to be fetch by
+	/// the caller.
+	/// For a child trie it is its root.
+	pub fn root(&self) -> Option<&[u8]> {
+		match self {
+			ChildInfo::Default(ChildTrie {
+				root,
+				data,
+			}) => if *root > 0 {
+				Some(&data[..*root])
+			} else {
+				None
+			}
+		}
+	}
+
+}
+
+/// Type of child.
+/// It does not strictly define different child type, it can also
+/// be related to technical consideration or api variant.
+#[repr(u32)]
+pub enum ChildType {
+	/// Default, it uses a cryptographic strong unique id as input.
+	CryptoUniqueId = 1,
+	/// Default, but with a root registerd to skip root fetch when querying.
+	/// Root is a 32 byte hash.
+	CryptoUniqueIdRoot32Api = 2,
```

Comment:
Does having the thing only implemented for root hash being 32 byte to too limiting ? don't we end to something a bit similar to what does fix https://github.com/paritytech/substrate/pull/4231

Reply:
We could add the hash length before the root in data? A bit extra memory but it would avoid consuming new child_type id.
I think it may be fine (since the root variant is no longer buildable from slice but only from a owned child info), I will implement it.